packages/webpack/src/builder.js

```diff
@@ -0,0 +1,284 @@
+
+    await runner(this.compilers, (compiler) => {
+      return this.webpackCompile(compiler)
+    })
+  }
+
+  webpackCompile(compiler) {
+    return new Promise(async (resolve, reject) => {
+      const name = compiler.options.name
+
+      await this.nuxt.callHook('build:compile', { name, compiler })
+
+      // Load renderer resources after build
+      compiler.hooks.done.tap('load-resources', async (stats) => {
+        await this.nuxt.callHook('build:compiled', {
+          name,
+          compiler,
+          stats
+        })
+
+        // Reload renderer if available
+        this.nuxt.renderer.loadResources(this.mfs || fs)
+
+        // Resolve on next tick
+        process.nextTick(resolve)
+      })
+
+      if (this.options.dev) {
+        // --- Dev Build ---
+        // Client Build, watch is started by dev-middleware
+        if (compiler.options.name === 'client') {
+          return this.webpackDev(compiler)
+        }
+        // Server, build and watch for changes
+        this.compilersWatching.push(
+          compiler.watch(this.options.watchers.webpack, (err) => {
+            /* istanbul ignore if */
+            if (err) return reject(err)
+          })
+        )
+      } else {
+        // --- Production Build ---
+        compiler.run((err, stats) => {
+          /* istanbul ignore next */
+          if (err) {
+            return reject(err)
+          } else if (stats.hasErrors()) {
+            if (this.options.build.quiet === true) {
+              err = stats.toString(this.options.build.stats)
+            }
+            if (!err) {
+              // actual errors will be printed by webpack itself
+              err = 'Nuxt Build Error'
+            }
+
+            return reject(err)
+          }
+
+          resolve()
+        })
+      }
+    })
+  }
+
+  webpackDev(compiler) {
+    consola.debug('Adding webpack middleware...')
+
+    // Create webpack dev middleware
+    this.webpackDevMiddleware = pify(
+      webpackDevMiddleware(
+        compiler,
+        Object.assign(
+          {
+            publicPath: this.options.build.publicPath,
+            stats: false,
+            logLevel: 'silent',
+            watchOptions: this.options.watchers.webpack
+          },
+          this.options.build.devMiddleware
+        )
+      )
+    )
+
+    this.webpackDevMiddleware.close = pify(this.webpackDevMiddleware.close)
+
+    this.webpackHotMiddleware = pify(
+      webpackHotMiddleware(
+        compiler,
+        Object.assign(
+          {
+            log: false,
+            heartbeat: 10000
+          },
+          this.options.build.hotMiddleware
+        )
+      )
+    )
+
+    // Inject to renderer instance
+    if (this.nuxt.renderer) {
+      this.nuxt.renderer.webpackDevMiddleware = this.webpackDevMiddleware
+      this.nuxt.renderer.webpackHotMiddleware = this.webpackHotMiddleware
+    }
+
+    // Start watching client files
+    this.watchClient()
+  }
+
+  watchClient() {
```

Comment:
Watch utilities are completely decoupled from webpack. What's your idea to move them into base builder? So we can reuse it for other builder :) @clarkdo

Reply:
Opps, it was in base builder, my mistake.